File: 2019/day10/day10.py
import math
import sys

def dist(a,b):
   return math.sqrt( (a[0] - b[0]) * (a[0] - b[0]) + (a[1] - b[1]) * (a[1] - b[1]) )

def sub(a,b):
    return [a[0] - b[0], a[1] - b[1]]

def angle2(a, b):
  return math.atan2(b[1] - a[1], b[0] - a[0])

def part1(a, pairs):
  lookup = []
  for t in pairs:
    u1 = sub(t[1], a)
    x = angle2([0,0], u1)
    md = math.degrees(x)
    md = round(md,10)
    if not md in lookup:
      lookup.append(md)
  return lookup

def part2(a, pairs):
  lookup = []
  for t in pairs:
    u1 = sub(t[1], a)
    x = angle2(a, t[1])
    md = math.degrees(x)
    md = round(md,10)
    lookup.append([t[1][0],t[1][1],md])
  return lookup

def los(m):
  rv = []
  for x in range(0, len(m[0])):
    for y in range(0, len(m)):
      if m[y][x] == ".":
        continue
      #print("-------------------")
      #print("#POINT",x,y)
      ck = []
      for x1 in range(0, len(m[0])):
        for y1 in range(0, len(m)):
          if m[y1][x1] == ".":
            continue
          if x == x1 and y == y1:
            continue
          ck.append([[x,y],[x1,y1]])
      angles = part1([x,y], ck)
      #print(sorted(angles))
      #print("")
      rv.append([x,y,len(angles)])
  mx = 0
  bs = ''
  for r in rv:
    #print("[",r[0],",",r[1],"]:",r[2])
    if r[2] > mx:
      bs = r[:2]
      mx = r[2]
  print("")
  print("Part 1:",bs,"::",mx)
  print("")

  #part2
  ck = []
  for x1 in range(0, len(m[0])):
    for y1 in range(0, len(m)):
      if m[y1][x1] == ".":
        continue
      if x == x1 and y == y1:
        continue
      ck.append([[x,y],[x1,y1]])
  aa = part2(bs, ck)
  for i in range(0, len(aa)):
     aa[i] = [aa[i][0],aa[i][1],(aa[i][2]+90+360)%360]
  shot = []
  lang = 0
  print("left:",len(aa),"shot:",len(shot))
  print("")

  def m(x):
    rv = ((x[2] - lang)%360) * 1000 + dist(bs, x)
    return rv

  while aa:
    if len(shot) >= 200:
      break
    tgt = min(aa, key=m)
    print("Target:",tgt,"m:", m(tgt),"a:","l:", lang, dist(bs, tgt))
    shot.append(tgt)
    aa.remove(tgt)
    lang = (tgt[2] + 0.001)%360
    print("left:",len(aa),"shot:",len(shot),"last:",tgt,"lang:",lang)
    print("")

  print("Part 1:",bs,"::",mx)
  print("Part 2:",tgt,"::",(tgt[0]*100+tgt[1]))

if __name__ == "__main__":
  if len(sys.argv) < 2:
    name = "../input/day10/part1"
  else:
    name = sys.argv[1]
  with open(name, "r") as fh:
    lines = fh.readlines()
    am = []
    for i in range(0, len(lines)):
        lines[i] = lines[i].strip()
        if not i in am:
            am.append([])
        for c in lines[i]:
            am[i].append(c)
  print(lines, len(lines[0]), "x", len(lines))
  los(am)

File: 2019/day10/day10.jl
# Julia 1.3.0

function dist(a, b)
	return sqrt( (a[1] - b[1]) * (a[1] - b[1]) + (a[2] - b[2]) * (a[2] - b[2]) )
end

function sub(a,b)
	return (a[1] - b[1], a[2] - b[2])
end

function angle2(a, b)
	return atan(b[2] - a[2], b[1] - a[1])
end

function part1(coord, pairs)
	rv = []
	for t in pairs
		u1 = sub(t[2], coord)
		x = angle2((0,0), u1)
		md = rad2deg(x)
		if !(md in rv)
			append!(rv, md)
		end
	end
	return rv
end

function part2(coord, pairs)
	rv = []
	for t in pairs
		u1 = sub(t[2], coord)
		x = angle2(coord, t[2])
		md = rad2deg(x)
		push!(rv, (t[2][1], t[2][2], md))
	end
	return rv
end

function los1(m)
	rv = []
	for y in 1:length(m)
		for x in 1:length(m[y])
			if m[y][x] == '.'
				continue
			end
			ck = []
			for y1 in 1:length(m)
				for x1 in 1:length(m[y])
					if m[y1][x1] == '.'
						continue
					end
					if x == x1 && y == y1
						continue
					end
					push!(ck, ((x,y),(x1,y1)))
				end
			end
			angles = part1((x,y), ck)
			push!(rv, (x,y,length(angles)))
		end
	end
	mx = 0
	bs = (0,0)
	for r in rv
		if r[3] > mx
			bs = (r[1], r[2])
			mx = r[3]
		end
	end
	println()
	println("Part 1: ($(bs[1]-1),$(bs[2]-1)) :: $(mx)")
	println()
	return (bs, mx)
end

function m(x, bs, lang)
	rv = ((((x[3] - lang) % 360) + 360) % 360) * 1000 + dist(bs, x)
	#println("    m(",x," ",bs," ",lang," ",dist(bs,x),")=",rv)
	return rv
end

function los2(mm, bs)
	ck = []
	for y1 in 1:length(mm)
		for x1 in 1:length(mm[y1])
			if mm[y1][x1] == '.'
				continue
			end
			if y1 == bs[2] && x1 == bs[1]
				continue
			end
			push!(ck, ((0,0),(x1,y1)))
		end
	end
	aa = part2(bs, ck)
	for i in 1:length(aa)
		aa[i] = (aa[i][1], aa[i][2], (aa[i][3] + 90 + 360) % 360)
	end
	for a in sort(aa)
		#println(a," ",a[3] * 10000 + dist(bs, a))
	end
	shot = []
	lang = 0
	println("left: ",length(aa)," shot: ",length(shot))
	println()

	tgt = (0,0,0)
	lang = 0
	while true
		if length(aa) < 1
			break
		end
		if length(shot) >= 200
			break
		end
		tmpa = sort(aa, by = x -> m(x, bs, lang))
		tgt = tmpa[1]
		idx = 1
		hx = m(tgt, bs, lang)
		println("Target: (",tgt[1],", ",tgt[2],", ",tgt[3],") m: ",hx ," l: ", lang)
		aa = deleteat!(tmpa, idx)
		push!(shot, tgt)
		lang = (tgt[3] + 0.001) % 360
		println("left: ", length(aa), " shot: ", length(shot), " last: ", tgt, " lang: ", lang)
		println()
	end
	println("Part 2: (", tgt[1]-1,"/",tgt[2]-1,") :: ", (tgt[1]-1)*100 + (tgt[2]-1))
	return tgt
end

filename = "../input/day10/part1"

if length(ARGS) > 0
	filename = ARGS[1]
end

println("File: $(filename)")
am = readlines(filename)
(bs, mx) = los1(am)
tgt = los2(am, bs)
println("---")
println("Part 1: ($(bs[1]-1),$(bs[2]-1)) :: $(mx)")
println("Part 2: (", tgt[1]-1,"/",tgt[2]-1,") :: ", (tgt[1]-1)*100 + (tgt[2]-1))
